### PR TITLE
migration: Move from vanilla JS to Next.JS

### PR DESCRIPTION
**Pull request by a1icja**
**tl;dr** - The dashboard has been rebuilt using Next.js, TailwindCSS, and PrimeReact (components library) to provide a better overall user and developer experience.

---

This PR reimplements the features found in the dashboard using Next.js. It is not intended to be used for adding new features or fixing old bugs.

Please see the `README.md` file included in this PR for a general overview of how the project is formatted as well as information for how to setup a development environment.

*A live version of this dashboard with additional improvements and features (coming in future PRs) can be found [here](https://a1icja.github.io/kata-dashboard-next/).*

---

In order for this PR to be merged successfully, a couple things must happen:

1. The [GitHub Pages settings for this repository](https://github.com/kata-containers/kata-containers.github.io/settings/pages) must be changed to [use GitHub Actions as the source.](https://github.com/user-attachments/assets/7d89c131-f337-4fea-bb79-cc3f1bb7b8f7)
2. The "Fetch CI Nightly Data" action must be run manually **once** *following* this PR being merged

---

Due to the significant changes present in this PR, please review each file carefully.

### CODE REVIEW

File: .github/workflows/fetch-ci-nightly-data.yaml [hunk 1]
```diff
@@ -13,17 +18,19 @@ jobs:
...
         run: |
           # fetch ci nightly data as temporary file
-          node js/fetch-ci-nightly-data.js | tee tmp-data.js
+          node scripts/fetch-ci-nightly-data.js | tee tmp-data.json
           # switch to a branch specifically for holding latest data
           git config --global user.name "GH Actions Workflow"
           git config --global user.email "<gha@runner>"
           git fetch --all
           git checkout latest-dashboard-data
-          # back out whatever data was there
```
Comment (afinn12) [inline .github/workflows/fetch-ci-nightly-data.yaml:22]: Might want to leave in.

Comment (a1icja) [reply .github/workflows/fetch-ci-nightly-data.yaml:22]: Could go either way. Anna and I talked, and it would be helpful to leave in to reduce the number of commits, but there is the case where it could break on a newly created branch. Might be helpful to do a check and run only if commits exist. Food for thought

File: scripts/fetch-ci-nightly-data.js [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+//     1 for the last 10 closed pull requests
+//    20 for all 4 batches of checks (max of a hundred each) for each of the 5 PRs
+//     X Other?
+// TOTAL: 43?
+// LIMIT: 60 per hour  // curl https://api.github.com/rate_limit
+// TODO: Further explore using the GraphQL API, which permits more narrowly targeted queries 
+
+// Github API URL for the kata-container ci-nightly workflow's runs. This
+// will only get the most recent 10 runs ('page' is empty, and 'per_page=10').
+var ci_nightly_runs_url =
```
Comment (Redent0r) [inline scripts/fetch-ci-nightly-data.js:27]: Why not use `const` for all these?

Comment (a1icja) [reply scripts/fetch-ci-nightly-data.js:27]: Definitely a change we can make! The use of `var` came from the current version of the script.

File: pages/index.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+      data = await response.json();
+
+      try {
+        const jobData = Object.keys(data).map((key) => {
+          const job = data[key];
+          return { name: key, ...job };
+        });
+        setJobs(jobData);
+      } catch (error) {
+        console.error("Error fetching data:", error);
```
Comment (Redent0r) [inline pages/index.js:37]: May consider showing some kind of toast or popup for the user in case they don't check the console

File: pages/index.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  }, []);
+
+  useEffect(() => {
+    setLoading(true);
+
+    // Create rows to set into table.
+    const rows = jobs.map((job) => ({
+      name: job.name,
+      runs: job.runs,
+      fails: job.fails,
```
Comment (arc9693) [inline pages/index.js:53]: Might wanna use [spread operator](https://www.w3schools.com/react/react_es6_spread.asp) to avoid rewriting assignments.

File: pages/index.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        "https://raw.githubusercontent.com/kata-containers/kata-containers.github.io" +
+          "/refs/heads/latest-dashboard-data/data/job_stats.json"
+      );
+      data = await response.json();
+
+      try {
+        const jobData = Object.keys(data).map((key) => {
+          const job = data[key];
+          return { name: key, ...job };
+        });
```
Comment (arc9693) [inline pages/index.js:34]: This logic can be moved out of this file. Could use a const var for url.

File: pages/index.js [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        <Image
+          src={`${basePath}/${icon}`}
+          alt="weather"
+          width={32}
+          height={32}
+          // priority
+        />
+      </div>
+    );
+  };
```
Comment (arc9693) [inline pages/index.js:103]: This could become another react component?

File: pages/index.js (hunk 5 above)
Comment (a1icja) [reply pages/index.js:34]: Definitely can move this out, but I don't quite follow the const var idea. Could you elaborate?

File: scripts/fetch-ci-nightly-data.js (hunk 2 above)
Comment (a1icja) [reply scripts/fetch-ci-nightly-data.js]: Changed in [e9c0fcb](https://github.com/kata-containers/kata-containers.github.io/pull/16/commits/e9c0fcbe6fff61b17c6cd8e4b9456bcbeae09f06)

File: pages/index.js (hunk 3 above)
Comment (a1icja) [reply pages/index.js:30]: Added a TODO comment for this since this PR isn't meant to add new features. When @sprt gets back, we can add it to the project board.

File: pages/index.js (hunk 4 above)
Comment (a1icja) [reply pages/index.js]: Changed in [5ed8858](https://github.com/kata-containers/kata-containers.github.io/pull/16/commits/5ed8858ca27efc8288386a941d4aa817aad7e7d4)

File: pages/index.js (hunk 6 above)
Comment (a1icja) [reply pages/index.js]: Changed in [f61c5df](https://github.com/kata-containers/kata-containers.github.io/pull/16/commits/f61c5dfd1883336588c366e8990a4abf2f75fcc1)

File: .github/workflows/fetch-ci-nightly-data.yaml (hunk 1 above)
Comment (sprt) [reply .github/workflows/fetch-ci-nightly-data.yaml:22]: Let's revert this so we minimize the number of changes we make and reduce the risk of breaking something.

File: .github/workflows/fetch-ci-nightly-data.yaml [hunk 7]
```diff
@@ -13,17 +18,19 @@ jobs:
...
-          git reset HEAD~1
+          # pull in the latest changes
+          git pull
           # overwrite the old data
+          git rm -r --cached data/
+          rm -rf data/
```
Comment (sprt) [inline .github/workflows/fetch-ci-nightly-data.yaml:31]: In this redundant if we already call `git rm`? We could also revert these 2 new lines to keep things simple.

File: scripts/fetch-ci-nightly-data.js [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+  }
+
+  // Fetch the checks for a PR.
+  function fetch_checks(p) {
+    if (p > 60) {
+      throw new Error(`Attempting to make too many API calls: ${p}`)
+    }
+    return fetch_checks_by_page(p)
+    .then(function (checks_request) {
+      // console.log('checks_request', checks_request);
```
Comment (sprt) [inline scripts/fetch-ci-nightly-data.js:181]:
```suggestion
```

File: scripts/fetch-ci-nightly-data.js [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+      throw error;
+    });
+  }
+
+  // Extract list of objects with PR commit SHAs, PR URLs, and PR number (i.e. id)
+  const prs_with_check_data = {
+    html_url: pr["html_url"],  // link to PR page
+    number: pr["number"],  // PR number (used as PR id); displayed on dashboard
+    commit_sha: pr["head"]["sha"],  // For getting checks run on PR branch
+    // commit_sha: pr["merge_commit_sha"],  // For getting checks run on main branch after merge
```
Comment (sprt) [inline scripts/fetch-ci-nightly-data.js:207]:
```suggestion
```

File: scripts/fetch-ci-nightly-data.js [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    checks: [],  // will be populated later with fetch_checks
+  };
+
+  return fetch_checks(1);
+}
+
+// Extract list of required jobs (i.e. main branch details: protection: required_status_checks: contexts)
+function get_required_jobs(main_branch) {
+  const required_jobs = main_branch["protection"]["required_status_checks"]["contexts"];
+  // console.log('required jobs: ', required_jobs);
```
Comment (sprt) [inline scripts/fetch-ci-nightly-data.js:218]:
```suggestion
```

File: .github/workflows/deploy.yml [hunk 11]
```diff
@@ -0,0 +1,98 @@
+# Sample workflow for building and deploying a Next.js site to GitHub Pages
```
Comment (sprt) [inline .github/workflows/deploy.yml:1]: Adding a link to the original for reference.
```suggestion
# Reference: https://github.com/actions/starter-workflows/blob/main/pages/nextjs.yml

# Sample workflow for building and deploying a Next.js site to GitHub Pages
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+---
+
+## Setup Instructions
+
+Follow these steps to set up the development environment for the Kata Containers Test Dashboard:
+
+### Prerequisites
+- **Node.js** (version 18.x or later recommended)
+- **npm** (comes with Node.js)
```
Comment (sprt) [inline README.md:54]: nit: Could we link to installation instructions here?

File: README.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+   npm start
+   ```
+
+### Scripts
+
+- **Fetch CI Nightly Data**: 
+  The `fetch-ci-nightly-data.js` script can be executed manually to pull the latest CI test data from the Kata Containers repository:
+   ```bash
+   node scripts/fetch-ci-nightly-data.js
+   ```
```
Comment (sprt) [inline README.md:100]: Could we add full instructions on how to test this e2e locally? If I run that command, it simply prints the API results to the console, but the dashboard is still empty. We need to have instructions so that anyone can set up a working dashboard locally and iterate.

**Review by sprt — CHANGES_REQUESTED** [review]
Thank you @a1icja and team! I left a few comments.

One ask - if you scroll down to the bottom of the PR page, you'll find that we're failing the [commit message check](https://github.com/kata-containers/kata-containers.github.io/actions/runs/11541292791/job/32175373802?pr=16). Can you follow the [contributing guidelines](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format) (the whole section up until *Reviews*, not just the first heading) and fix your PR? TLDR: Squash your commits into one, update the commit subject to match the new PR title, add a description and a signoff.

One question - do you know how hard it would be to be able to set up a staging website per PR to be able to test PRs publicly before merging them? Otherwise folks would have to manually checkout feature branches to test changes.

File: .github/workflows/fetch-ci-nightly-data.yaml (hunk 7 above)
Comment (a1icja) [reply .github/workflows/fetch-ci-nightly-data.yaml:28]: It's "redundant", but I had to add it in order to ensure that the data is always replaced on update.

File: README.md (hunk 12 above)
Comment (a1icja) [reply README.md:54]: Can do! I'm partial to [fnm](https://github.com/Schniz/fnm), but I can also do the official instructions if you'd like.

Comment (sprt) [reply README.md:54]: We can just link to https://nodejs.org/en/download - folks can choose the installer they prefer

File: .github/workflows/fetch-ci-nightly-data.yaml (hunk 7 above)
Comment (sprt) [reply .github/workflows/fetch-ci-nightly-data.yaml:28]: Could you explain further?

File: README.md (hunk 12 above)
Comment (a1icja) [reply README.md]: Implemented

File: README.md (hunk 13 above)
Comment (a1icja) [reply README.md]: Instructions updated

File: .github/workflows/fetch-ci-nightly-data.yaml (hunk 7 above)
Comment (a1icja) [reply .github/workflows/fetch-ci-nightly-data.yaml]: I removed/reverted the `rm` command because removing `--cached` also deletes off disk.

File: pages/index.js [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+                    {emoji} {run.run_num}
+                  </a>
+                  &nbsp;&nbsp;&nbsp;&nbsp;
+                </span>
+              );
+            })
+          ) : (
+            <div>No Nightly Runs associated with this job</div>
+          )}
+        </div>``
```
Comment (sprt) [inline pages/index.js:118]:
```suggestion
        </div>
```

File: .github/workflows/fetch-ci-nightly-data.yaml [hunk 15]
```diff
@@ -13,17 +15,20 @@ jobs:
...
           git config --global user.name "GH Actions Workflow"
           git config --global user.email "<gha@runner>"
           git fetch --all
           git checkout latest-dashboard-data
           # back out whatever data was there
           git reset HEAD~1
+          # pull in the latest changes
+          git pull
           # overwrite the old data
+          git rm -r data/
```
Comment (sprt) [inline .github/workflows/fetch-ci-nightly-data.yaml:29]: Let's revert to minimize the amount of changes, unless we have a strong reason to do this
```suggestion
          # overwrite the old data
```

**Review by sprt — CHANGES_REQUESTED** [review]
Just a couple more comments. Very close, thanks!
